Add support for categorical feature auto-encoding

nunosilva800 opened the pull request:
Adds support to have categorical features label encoded automatically, when using Booster#predict.

The python api does it by relying on pandas dataframes: https://github.com/microsoft/LightGBM/blob/e057ae08e6bf6c6c84f276a127423fb145ca5fdb/python-package/lightgbm/basic.py#L1131-L1137

Would be great for this ruby lib to provide similar quality-of-life.

ankane commented:
Hi @nunosilva800, thanks for the another PR. Happy to include support for this, but it'd be good if the code matched Python as closely as possible (rather than a separate class).

Also, please share the code to generate the model file rather than checking it in directly.

nunosilva800 commented:
@ankane I've updated files in test/support to match the lightgbm v4 python API and added the categorical model file.

> good if the code matched Python as closely as possible (rather than a separate class).

I'd like to challenge that. The python package overrides the input `data` when its type is a pandas dataframe. The whole logic is inside `_data_from_pandas`, but is stateless (does not mutate `self`) and therefore could be very well extracted somewhere else. 

Here in ruby land, there is no pandas dataframe (apart from maybe [mrkn/pandas.rb](https://github.com/mrkn/pandas.rb)), so the approach here is to always transform the categorical features, if their names are defined as categorical in the model.
My thinking in extracting into a separate class is twofold:
1. It is clearer to navigate the code and figure out what is happening if this transformation is causing predictions to be unexpected
2. If in the future it turns out to be better to let the caller decide when to apply this transformation, they can do so by calling `LightGBM::CategoricalFeatureEncoder` directly, before invoking `LightGBM::Booster#predict`


What do you think?

ankane commented:
Added a version of this in the commit above that matches the Python code a bit better.
